daemons/ipa-slapi-plugins/ipa-pwd-extop/prepost.c

```diff
@@ -1520,8 +1545,48 @@ static int ipapwd_pre_bind(Slapi_PBlock *pb)
     /* Attempt to write out kerberos keys for the user. */
     ipapwd_write_krb_keys(pb, discard_const(dn), entry, credentials);
 
+    /* Get the (user's) uid and $SUFFIX parts of the DN. */
+    uid_len = (size_t)(strchr(dn, ',') - dn);
+    strncpy(uid, dn, uid_len);
+    uid[uid_len] = '\0';
+    suffix_start = strstr(dn, ",dc=");
+    suffix_len = strlen(dn) - (size_t)(suffix_start - dn);
+    suffix = malloc(suffix_len + 1);
+    strncpy(suffix, suffix_start, suffix_len);
+
+    /* Also compare against app passwords of the user. If some of them matches,
+     * the user is authenticated using that password and he/she will have
+     * read-only access to his primary entry. No operations with OTP or
+     * Kerberos are performed for app passwords.
+     */
+    for (i = 0; i < APPPW_MAX_COUNT; i++) {
+        sprintf(apppw_dn, "%d,%s,%s%s", i, uid, APPPW_CONTAINER_DN, suffix);
```

Comment:
slapi_ch_smprintf() or asprintf() is preferred. and check that an OOM didn't happen.

Reply:
I decided to use slapi_ch_smprintf(). I now check for OOM throughout entire new code.